akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala

```diff
@@ -915,8 +913,36 @@ trait FlowOps[+Out, +Mat] {
    * '''Cancels when''' downstream cancels and all substreams cancel
    *
    */
-  def groupBy[K, U >: Out](f: Out ⇒ K): Repr[(K, Source[U, Unit]), Mat] =
-    deprecatedAndThen(GroupBy(f.asInstanceOf[Any ⇒ Any]))
+  def groupBy[K](f: Out ⇒ K): SubFlow[Out, Mat, Repr, Closed] = {
+    implicit def mat = GraphInterpreter.currentInterpreter.materializer
+    def makeEager[T]: Graph[FlowShape[T, T], Unit] = new GraphStage[FlowShape[T, T]] {
+      val in = Inlet[T]("in")
+      val out = Outlet[T]("out")
+      override val shape = FlowShape(in, out)
+      override def createLogic(attr: Attributes) = new GraphStageLogic(shape) {
+        passAlong(in, out, doFinish = true, doFail = true)
+        setHandler(out, eagerTerminateOutput)
+        override def preStart(): Unit = pull(in)
+      }
+    }
+    val merge = new SubFlowImpl.MergeBack[Out, Repr] {
+      override def apply[T](flow: Flow[Out, T, Unit], breadth: Int): Repr[T] =
+        deprecatedAndThen[Source[Out, Unit]](GroupBy(f.asInstanceOf[Any ⇒ Any]))
+          .map(_.via(flow).via(makeEager) /* needed to start generating some demand */ )
+          .map(s ⇒ Source(s.runWith(Sink.publisher(false)))) // need to be eager as a workaround for now
+          .buffer(4, OverflowStrategy.Backpressure) // needed to make reduceByKey recipe work
```

Comment:
Ah, ok, then if it is how many streams are merged in concurrently I think it would be nice to have a cohesive name for that parameter value, as it is related to mapAsync/mapAsyncUnordered etc. so mergeBack(1) == concat